apps/android/LiveVideoApp/app/src/main/java/com/twilio/livevideo/app/manager/GridManager.kt

```diff
@@ -0,0 +1,153 @@
+            //New participants To Be Added
+            var limit = GRID_MAX_PARTICIPANTS_LIMIT - participants.size
+            if (limit < 0) limit = 0
+            participantsToBeAdded.take(limit).forEach { item ->
+                val participantViewBinding: ParticipantViewItemBinding =
+                    createParticipantView(inflater, item, gridLayout, isVerticalModeEnabled, lifecycleOwner)
+                participants.add(item)
+                bindings.add(participantViewBinding)
+            }
+        }
+    }
+
+    private fun createParticipantView(
+        inflater: LayoutInflater,
+        item: ParticipantStream,
+        parentView: ViewGroup?,
+        isVerticalModeEnabled: Boolean,
+        lifecycleOwner: LifecycleOwner,
+    ): ParticipantViewItemBinding {
+
+        val participantViewBinding = ParticipantViewItemBinding.inflate(inflater, parentView, false)
+        participantViewBinding.item = item
+        participantViewBinding.lifecycleOwner = lifecycleOwner
+
+        participantViewBinding.root.doOnAttach {
+            Timber.d("doOnAttach ${item.identity}")
+            it.doOnDetach {
+                Timber.d("doOnDetach ${item.identity}")
+                item.videoTrack?.removeSink(participantViewBinding.videoView)
+                lifecycleOwner.lifecycle.removeObserver(item)
+            }
+        }
+
+        parentView?.apply {
+            addView(
+                participantViewBinding.root,
+                participantViewBinding.root.layoutParams.apply {
+                    toggleGridMode((this as GridLayout.LayoutParams), isVerticalModeEnabled)
+                }
+            )
+        }
+
+        return participantViewBinding
+    }
+
+    private fun toggleGridMode(layoutParams: GridLayout.LayoutParams, isVerticalModeEnabled: Boolean) {
+        layoutParams.columnSpec = if (isVerticalModeEnabled) {
+            GridLayout.spec(GridLayout.UNDEFINED, 2, 1F)
+        } else {
+            GridLayout.spec(GridLayout.UNDEFINED, 1, 1F)
+        }
+    }
+
+    companion object {
+        private const val GRID_MAX_PARTICIPANTS_LIMIT = 6
+        private const val VERTICAL_MODE_LIMIT = 3
+
+        fun switchParticipants(list: MutableList<ParticipantStream>, participant: ParticipantStream): Boolean {
+            list.indexOf(participant).also { index ->
+                if (index + 1 > GRID_MAX_PARTICIPANTS_LIMIT) {
+                    //Take the first N limit participants and skip the first element which is the local participant.
+                    list.take(GRID_MAX_PARTICIPANTS_LIMIT).drop(1).minByOrNull {
+                        it.dominantSpeakerStartTime
+                    }?.also { oldestDominantSpeaker ->
+                        val oldestDominantSpeakerIndex = list.indexOf(oldestDominantSpeaker)
+                        list[oldestDominantSpeakerIndex] = participant
+                        list[index] = oldestDominantSpeaker
```

Comment:
Looks like this code is swapping participants. iOS and web apps behave differently. When the oldest dominant speaker is removed from the grid, they are always moved to the start of the offscreen list. This way if someone on the grid disconnects later, the most recent dominant speaker off screen will be moved offscreen.

See how this is a 4 step process on iOS: https://github.com/twilio/twilio-live-interactive-video/blob/eba1c8a78235cc22a90cf2bfd3780d6b267dfbf3/apps/ios/LiveVideo/LiveVideo/Views/Stream/SpeakerGridViewModel.swift#L110-L113

Reply:
Also it seems like `GridManager` should not modify internal state of `RoomManager`. It gets to be difficult to follow all the places the state is being modified.

It seems like the grid state is split between `GridManager` and `RoomManager`. I would recommend having `GridManager` own state for all participants regardless if they are visible or not on the grid. 

Eventually you will be implementing speaker layout which is a completely different layout. The User will be able to switch between grid and speaker layouts. It is important for each layout to manage their own state so that one layout has no impact on another layout.

Reply:
Done!